src/ATen/native/xpu/mkl/SpectralOps.cpp

```diff
@@ -0,0 +1,461 @@
+  signal_size[0] = batch_size;
+
+  for (int64_t i = 0; i < signal_ndim; ++i) {
+    auto in_size = input.sizes()[i + 1];
+    auto out_size = out_sizes[dim[i]];
+    signal_size[i + 1] = std::max(in_size, out_size);
+    TORCH_INTERNAL_ASSERT(
+        in_size == signal_size[i + 1] ||
+        in_size == (signal_size[i + 1] / 2) + 1);
+    TORCH_INTERNAL_ASSERT(
+        out_size == signal_size[i + 1] ||
+        out_size == (signal_size[i + 1] / 2) + 1);
+  }
+
+  batched_sizes[0] = batch_size;
+  DimVector batched_out_sizes(batched_sizes.begin(), batched_sizes.end());
+
+  for (size_t i = 0; i < dim.size(); ++i) {
+    batched_out_sizes[i + 1] = out_sizes[dim[i]];
+  }
+
+  out.resize_(batched_out_sizes, MemoryFormat::Contiguous);
+
+  // run the FFT
+  _fft_with_size(
+      out,
+      input,
+      signal_ndim,
+      input.is_complex(),
+      out.is_complex(),
+      !forward,
+      signal_size,
+      onesided);
+
+  // Inplace reshaping to original batch shape and inverting the dimension
+  // permutation
+  DimVector out_strides(ndim);
+  int64_t batch_numel = 1;
+
+  for (int64_t i = batch_dims - 1; i >= 0; --i) {
+    out_strides[dim_permute[i]] = batch_numel * out.strides()[0];
+    batch_numel *= out_sizes[dim_permute[i]];
+  }
+
+  for (int64_t i = batch_dims; i < ndim; ++i) {
+    out_strides[dim_permute[i]] = out.strides()[1 + (i - batch_dims)];
+  }
+
+  out.as_strided_(out_sizes, out_strides, out.storage_offset());
+
+  return out;
+}
+
+double _dft_scale(
+    IntArrayRef dim,
+    IntArrayRef input_sizes,
+    IntArrayRef out_sizes,
+    int64_t normalization) {
+  const auto norm = static_cast<fft_norm_mode>(normalization);
+  double double_scale = 1.0;
+
+  if (norm == fft_norm_mode::none) {
+    return double_scale;
+  }
+
+  const int64_t signal_ndim = dim.size();
+  int64_t signal_numel = 1;
+
+  for (int64_t i = 0; i < signal_ndim; ++i) {
+    auto in_size = input_sizes[dim[i]];
+    auto out_size = out_sizes[dim[i]];
+    auto signal_size = std::max(in_size, out_size);
+
+    signal_numel *= signal_size;
+    TORCH_INTERNAL_ASSERT(
+        in_size == signal_size || in_size == (signal_size / 2) + 1);
+    TORCH_INTERNAL_ASSERT(
+        out_size == signal_size || out_size == (signal_size / 2) + 1);
+  }
+
+  if (norm == fft_norm_mode::by_root_n) {
+    double_scale = 1.0 / std::sqrt(signal_numel);
+  } else {
+    double_scale = 1.0 / static_cast<double>(signal_numel);
+  }
+
+  return double_scale;
+}
+
+const Tensor& _fft_apply_normalization(
+    const Tensor& self,
+    int64_t normalization,
+    IntArrayRef sizes,
+    IntArrayRef dims) {
+  auto scale = _dft_scale(dims, sizes, self.sizes(), normalization);
+  return (scale == 1.0) ? self : self.mul_(scale);
+}
+
+} // namespace impl
+
+Tensor _fft_c2c_mkl(
+    const Tensor& self,
+    IntArrayRef dim,
+    int64_t normalization,
+    bool forward) {
+  TORCH_CHECK(self.is_complex());
+
+  if (dim.empty()) {
+    return self.clone();
+  }
+
```

Comment:
Duplicate. You can assume the they always meet requirement.

Reply:
The checking for empty dims is to fix `test_fftn_noop_transform`. Please refer to https://github.com/pytorch/pytorch/pull/117368 for details.

Reply:
Can you
1. remove them to op level?
or
2. remove `TORCH_CHECK(self.is_complex());` Since the check exists in op level already.

Reply:
`TORCH_CHECK(self.is_complex());` has been removed.